tensorboard/plugins/hparams/list_session_groups.py

```diff
@@ -0,0 +1,393 @@
+    # result is a list of api_pb2.MetricValue instances.
+    result=[]
+    metric_infos = self._context.experiment().metric_infos
+    for metric_info in metric_infos:
+      metric_name = metric_info.name
+      try:
+        metric_evals = metrics.list_metric_evals(
+            self._context.multiplexer(),
+            session_name,
+            metric_name)
+      except KeyError as e:
+        # It's ok if we don't find the metric in the session.
+        # We skip it here for filtering and sorting purposes its value is _NULL
+        continue
+      # metric_evals[i] is a 3-tuple of the form [wall_time, step, value]
+      result.append(api_pb2.MetricValue(name=metric_name,
+                                        wall_time_secs=metric_evals[-1][0],
+                                        training_step=metric_evals[-1][1],
+                                        value=metric_evals[-1][2]))
+    return result
+
+  def _build_group_from_sessions(self, sessions, session_infos_by_name):
+    assert len(sessions) > 0
+    # TODO(erez): Do proper metric aggregation. For now we just take
+    # the metric values from the first session.
+    result = api_pb2.SessionGroup(
+        name=session_infos_by_name[sessions[0].name].start_info.group_name,
+        metric_values=sessions[0].metric_values,
+        # Sort sessions by name so the order is deterministic.
+        sessions=sorted(sessions, key=lambda session : session.name),
+        monitor_url=(
+            session_infos_by_name[sessions[0].name].start_info.monitor_url
+        )
+    )
+    # Copy hparams from the first session (all sessions should have the same
+    # hyperparameter values) into result.
+    # There doesn't seem to be a way to initialize a protobuffer map in the
+    # constructor.
+    hparams = session_infos_by_name[sessions[0].name].start_info.hparams
+    for (key, value) in six.iteritems(hparams):
+      result.hparams[key].CopyFrom(value)
+    return result
+
+  def _filter(self, session_groups):
+    return filter(
+        lambda sg : self._passes_all_filters(sg),
+        session_groups)
+
+  def _passes_all_filters(self, session_group):
+    for f in self._filters:
+      if not f.passes(session_group):
+        return False
+    return True
+
+  def _sort(self, session_groups):
+    """Sorts 'session_groups' in place according to _request.col_params"""
+    # Sort by session_group name so we have a deterministic order.
+    session_groups.sort(key=lambda session_group : session_group.name)
+    # Sort by lexicographical order of the _request.col_params whose order
+    # is not ORDER_UNSPECIFIED. The first such column is the primary sorting
+    # key, the second is the secondary sorting key, etc. To achieve that we
+    # need to iterate on these columns in reverse order (thus the primary key
+    # is the key used in the last sort).
+    for col_param, extractor in reversed(zip(self._request.col_params,
```

Comment:
We can wait, sure.  Fair point that it's not obviously more efficient in complexity terms, but I expect it would be in practice due to smaller constant factors.  The column extractions all happen in either case, but with a single sort a given session group is only processed once to produce a slightly larger key, whereas with k sorts we have to retrieve the session group itself from memory potentially k times on separate passes.  I doubt key comparison would really be more expensive in typical cases, since tuple comparison is lexicographic and will stop as early as there's a difference, so it won't proceed past the first column that has arbitrarily finely distributed values (e.g. metric values or non-discrete hparam values).

Reply:
I suspect you're right. But I'd still like to see how important that optimization is with respect to other bottlenecks before doing it.